### DISCUSSION TITLE
ST7735 TFT LCD + Frame Buffer on Pi Pico

### DISCUSSION THREAD

**Discussion by HVW-3U** [micropython/micropython#16228] [Hardware & Peripherals]
Hi All
This is my first post to this discussion forum. Forgive me if I have not yet figured out the exact details of communicating here.
I am building a UI on a 1.8" ST7735  SPI TFT LCD with Pi Pico running MicroPython
System is working fine, but as reported everywhere, the ST7735 updating is annoyingly slow with the ST7735.py library addressing every pixel.
Even with minimizing the amount of data ( only changes) written to the display (re-write old value in background color, then write new value), the updates are quite visible.

I have been playing with  the Frame Buffer method.
The Bouncing Ball demo is blazing fast. I am getting smooth 70 FPS with SPI running stable @ 40Mhz.
But the MicroPython buildin Frame Buffer class seems to support only a very limited set of graphics functions.
From the ST7735.py lib I use the font selection (I need some special characters in the font) and the font size selection ( I need double size font for readability)

Library development is not my area of expertise, but it seems to me that in the ST7735 lib it should not be too difficult to write a pixel to the Frame buffer ByteArray instead of to the SPI interface. Dumping the FrameBuffer to the display seems to be no problem, even within the ST7735 lib.

Anybody aware of any work done in this area already ? (If yes, where to find ?)
All the demos that I have found up til now are limited to simple operations like drawing circles or boxes as supported by the FrameBuffer buildin lib.

**Comment by peterhinch** [comment]
A driver for ST7735r is [described here](https://github.com/peterhinch/micropython-nano-gui/blob/master/DRIVERS.md#31-drivers-for-st7735r). The driver works with [nano-gui](https://github.com/peterhinch/micropython-nano-gui) and other GUI libraries.

**Comment by bidrohini** [comment]
I think you may need to modify the ST7735 Library.

**Comment by rkompass** [comment]
Did you note the recently presented [Tempe](https://github.com/orgs/micropython/discussions/16105) graphics library?
IIUC it needs a graphics driver that is able to transfer a line of the framebuffer (i.e. an array of colours) to the display via PCI.
